File: diamond/src/diamond.erl
-module(diamond).

-export([rows/1]).

%% A = 65
space_out(L,Size) -> [L | lists:duplicate(Size, $\s)] ++ [L].

outer(Size) ->
    lists:duplicate(Size, $\s).

row($A,Width) ->
    Outer = outer(Width div 2),
    [ Outer ++ "A" ++ Outer ];
row(L, Width) ->
    Size = (L - 1 - $A) * 2 + 1,
    Outer = outer(( Width - Size - 2 ) div 2),
    Middle = space_out(L,Size),
    [ Outer ++ Middle ++ Outer ].

gen_rows($A , Width ) ->
    row($A,Width);
gen_rows(L, Width) ->
    gen_rows(L-1, Width) ++ row(L,Width).

rows("A") ->
    [ "A" ];
rows([L]) ->
    Width = (L - $A) * 2 + 1,
    First = gen_rows(L-1,Width), 
    First ++ row(L,Width) ++ lists:reverse(First).

File: diamond/lib/diamond.ex
defmodule Diamond do
  @doc """
  Given a letter, it prints a diamond starting with 'A',
  with the supplied letter at the widest point.
  """
  @spec build_shape(char) :: String.t()
  def build_shape(?A) do
    "A\n"
  end

  def build_shape(char) do
    width = (char - ?A) * 2 + 1
    first = gen_rows(char - 1, width)
    (first ++ row(char, width) ++ Enum.reverse(first)) |> String.Chars.to_string()
  end

  defp gen_rows(?A, width) do
    row(?A, width)
  end

  defp gen_rows(char, width) do
    gen_rows(char - 1, width) ++ row(char, width)
  end

  defp row(?A, width) do
    outer = outer(div(width, 2))
    [outer ++ [?A] ++ outer ++ '\n']
  end

  defp row(char, width) do
    size = (char - 1 - ?A) * 2 + 1
    outer = outer(div(width - size - 2, 2))
    middle = space_out(char, size)
    [outer ++ middle ++ outer ++ '\n']
  end

  defp outer(size) do
    List.duplicate(?\s, size)
  end

  defp space_out(char, size) do
    [char | List.duplicate(?\s, size)] ++ [char]
  end
end
